## [Problem 3] Learning Iris (binary classification) with Keras

In [1]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load Iris dataset
iris = load_iris()
X = iris.data
y = iris.target

# Convert target to one-hot encoding for binary classification (Iris-versicolor vs Iris-virginica)
y_binary = np.where(y == 1, 1, 0).reshape(-1, 1)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y_binary, test_size=0.2, random_state=42)

# Define the model using Sequential API
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=50, batch_size=1, verbose=1)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f'Binary classification accuracy on test set: {accuracy}')

# Predictions
predictions = model.predict(X_test)


2024-07-02 14:18:11.935973: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-02 14:18:12.975941: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


Epoch 1/50


/home/topister/Desktop/dpro/machine-learning/dpro/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


120/120 ━━━━━━━━━━━━━━━━━━━━ 1s 976us/step - accuracy: 0.5761 - loss: 1.2020  
Epoch 2/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 837us/step - accuracy: 0.6717 - loss: 0.6590
Epoch 3/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 822us/step - accuracy: 0.7069 - loss: 0.6237
Epoch 4/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 817us/step - accuracy: 0.6889 - loss: 0.6266
Epoch 5/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 828us/step - accuracy: 0.6052 - loss: 0.6703
Epoch 6/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 825us/step - accuracy: 0.6923 - loss: 0.6096
Epoch 7/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 829us/step - accuracy: 0.5614 - loss: 0.6693 
Epoch 8/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 844us/step - accuracy: 0.4847 - loss: 0.6818 
Epoch 9/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 879us/step - accuracy: 0.6735 - loss: 0.6014
Epoch 10/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 815us/step - accuracy: 0.6168 - loss: 0.5993 
Epoch 11/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 848us/step - accuracy: 0.5823 - loss: 0.6322 
Epoch 12/50
120/120 ━━━━━━━━━━━━━━━

## [Problem 4] Learn Iris (multi-level classification) with Keras

In [3]:
# Convert target to one-hot encoding for multi-class classification (all three Iris species)
encoder = OneHotEncoder(sparse_output=False)
y_multi = encoder.fit_transform(y.reshape(-1, 1))

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y_multi, test_size=0.2, random_state=42)

# Define the model using Sequential API
model_multi = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(3, activation='softmax')  # 3 classes for Iris species
])

# Compile the model
model_multi.compile(optimizer='adam',
                    loss='categorical_crossentropy',
                    metrics=['accuracy'])

# Train the model
model_multi.fit(X_train, y_train, epochs=50, batch_size=1, verbose=1)

# Evaluate the model
loss_multi, accuracy_multi = model_multi.evaluate(X_test, y_test, verbose=0)
print(f'Multi-class classification accuracy on test set: {accuracy_multi}')

# Predictions
predictions_multi = model_multi.predict(X_test)


Epoch 1/50


/home/topister/Desktop/dpro/machine-learning/dpro/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


120/120 ━━━━━━━━━━━━━━━━━━━━ 1s 862us/step - accuracy: 0.3587 - loss: 6.5779   
Epoch 2/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 920us/step - accuracy: 0.3379 - loss: 2.2985
Epoch 3/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 825us/step - accuracy: 0.4670 - loss: 1.0731
Epoch 4/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 837us/step - accuracy: 0.5984 - loss: 0.9801 
Epoch 5/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 876us/step - accuracy: 0.7301 - loss: 0.8772
Epoch 6/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 918us/step - accuracy: 0.7858 - loss: 0.8258
Epoch 7/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 883us/step - accuracy: 0.7349 - loss: 0.7497
Epoch 8/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 876us/step - accuracy: 0.7295 - loss: 0.7487
Epoch 9/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 899us/step - accuracy: 0.6888 - loss: 0.6953
Epoch 10/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8330 - loss: 0.6105
Epoch 11/50
120/120 ━━━━━━━━━━━━━━━━━━━━ 0s 894us/step - accuracy: 0.7707 - loss: 0.6175
Epoch 12/50
120/120 ━━━━━━━━━━━━━━━━━━━

## [Problem 5] Learning House Prices with Keras

In [5]:
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

# Load Boston housing dataset
boston = fetch_openml(data_id=531)

# Extract features and target
X, y = boston.data, boston.target

# Normalize data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Define the model using Sequential API
model_house_prices = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(1)  # Output layer, no activation for regression
])

# Compile the model
model_house_prices.compile(optimizer='adam',
                           loss='mean_squared_error',  # MSE for regression
                           metrics=['mae'])  # Mean Absolute Error

# Train the model
model_house_prices.fit(X_train, y_train, epochs=50, batch_size=1, verbose=1)

# Evaluate the model
loss, mae = model_house_prices.evaluate(X_test, y_test, verbose=0)
print(f'Mean Absolute Error on test set: {mae}')

# Predictions
predictions = model_house_prices.predict(X_test)


Epoch 1/50


/home/topister/Desktop/dpro/machine-learning/dpro/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


404/404 ━━━━━━━━━━━━━━━━━━━━ 1s 937us/step - loss: 613.8762 - mae: 22.7914
Epoch 2/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 875us/step - loss: 369.4716 - mae: 17.4083
Epoch 3/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 772us/step - loss: 208.3295 - mae: 11.9724
Epoch 4/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 728us/step - loss: 148.8708 - mae: 9.5631
Epoch 5/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 753us/step - loss: 107.3173 - mae: 7.9883
Epoch 6/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 743us/step - loss: 77.9535 - mae: 6.1701
Epoch 7/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 748us/step - loss: 48.8088 - mae: 4.6816
Epoch 8/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 744us/step - loss: 39.7437 - mae: 4.3092
Epoch 9/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 738us/step - loss: 33.0469 - mae: 4.2322
Epoch 10/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 41.4438 - mae: 4.3243
Epoch 11/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 813us/step - loss: 24.6615 - mae: 3.8182
Epoch 12/50
404/404 ━━━━━━━━━━━━━━━━━━━━ 0s 757us/step - loss: 28.7869 - mae: 3

## [Problem 6] Learning MNIST with Keras

In [6]:
import tensorflow as tf

# Load and preprocess MNIST dataset
mnist = tf.keras.datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train, X_test = X_train / 255.0, X_test / 255.0  # Normalize pixel values to [0, 1]

# Define the model using Sequential API
model_mnist = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),  # Flatten 28x28 images to 784 vector
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),  # Dropout regularization for overfitting prevention
    tf.keras.layers.Dense(10, activation='softmax')  # Output layer with 10 classes (digits 0-9)
])

# Compile the model
model_mnist.compile(optimizer='adam',
                    loss='sparse_categorical_crossentropy',  # Sparse categorical cross-entropy for multi-class classification
                    metrics=['accuracy'])

# Train the model
model_mnist.fit(X_train, y_train, epochs=10, verbose=1)

# Evaluate the model
loss, accuracy = model_mnist.evaluate(X_test, y_test, verbose=0)
print(f'Accuracy on test set: {accuracy}')

# Predictions
predictions = model_mnist.predict(X_test)


/home/topister/Desktop/dpro/machine-learning/dpro/lib/python3.10/site-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10


2024-07-02 14:31:10.049970: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 188160000 exceeds 10% of free system memory.


1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8631 - loss: 0.4762
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9548 - loss: 0.1485
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9673 - loss: 0.1080
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9738 - loss: 0.0870
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9758 - loss: 0.0754
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9788 - loss: 0.0663
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9824 - loss: 0.0563
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9827 - loss: 0.0508
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9851 - loss: 0.0468
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9856 - loss: 0.0428
Accuracy on test set: 0.9815000295639038
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


## [Problem 7] (Advance assignment) Rewriting to PyTorch

#### Problem 5: Learning House Prices with PyTorch

In [7]:
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader

# Load Boston housing dataset
boston = fetch_openml(data_id=531)

# Extract features and target
X, y = boston.data, boston.target

# Normalize data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Convert data to PyTorch tensors
X_tensor = torch.tensor(X_scaled, dtype=torch.float32)
y_tensor = torch.tensor(y, dtype=torch.float32).view(-1, 1)  # Reshape y to (n_samples, 1)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_tensor, y_tensor, test_size=0.2, random_state=42)

# Define the model
class HousePricesModel(nn.Module):
    def __init__(self):
        super(HousePricesModel, self).__init__()
        self.fc1 = nn.Linear(X_train.shape[1], 10)
        self.fc2 = nn.Linear(10, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

# Instantiate the model, loss function, and optimizer
model_house_prices = HousePricesModel()
criterion = nn.MSELoss()  # Mean Squared Error loss for regression
optimizer = optim.Adam(model_house_prices.parameters(), lr=0.001)

# Create DataLoader for batch processing
train_dataset = TensorDataset(X_train, y_train)
train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)

# Training loop
for epoch in range(50):
    for inputs, labels in train_loader:
        optimizer.zero_grad()
        outputs = model_house_prices(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

# Evaluate the model
with torch.no_grad():
    outputs = model_house_prices(X_test)
    loss = criterion(outputs, y_test)
    mae = torch.mean(torch.abs(outputs - y_test))

print(f'Mean Absolute Error on test set: {mae.item()}')

# Predictions
predictions = outputs.detach().numpy()


Mean Absolute Error on test set: 2.389925956726074


### Problem 6: Learning MNIST with PyTorch

In [8]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision.transforms as transforms
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader

# Define transformations and load MNIST dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))
])
train_dataset = MNIST(root='./data', train=True, download=True, transform=transform)
test_dataset = MNIST(root='./data', train=False, download=True, transform=transform)

# Create DataLoader for batch processing
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

# Define the model
class MNISTModel(nn.Module):
    def __init__(self):
        super(MNISTModel, self).__init__()
        self.flatten = nn.Flatten()
        self.fc1 = nn.Linear(28*28, 128)
        self.dropout = nn.Dropout(0.2)
        self.fc2 = nn.Linear(128, 10)  # 10 classes for digits 0-9

    def forward(self, x):
        x = self.flatten(x)
        x = torch.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

# Instantiate the model, loss function, and optimizer
model_mnist = MNISTModel()
criterion = nn.CrossEntropyLoss()  # Cross-entropy loss for multi-class classification
optimizer = optim.Adam(model_mnist.parameters(), lr=0.001)

# Training loop
for epoch in range(10):
    model_mnist.train()
    for inputs, labels in train_loader:
        optimizer.zero_grad()
        outputs = model_mnist(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

# Evaluation
model_mnist.eval()
correct = 0
total = 0
with torch.no_grad():
    for inputs, labels in test_loader:
        outputs = model_mnist(inputs)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

accuracy = correct / total
print(f'Accuracy on test set: {accuracy}')


Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 9912422/9912422 [01:20<00:00, 123421.71it/s]


Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 28881/28881 [00:00<00:00, 122391.18it/s]


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 1648877/1648877 [00:10<00:00, 154100.19it/s]


Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 4542/4542 [00:00<00:00, 1888434.65it/s]


Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw

Accuracy on test set: 0.9708


## [Problem 8] (Advance assignment) Comparison of frameworks

Based on the results from Problems 3, 4, 5, 6, and in PyTorch for Problems 7, let's summarize the comparison between TensorFlow/Keras and PyTorch:

### TensorFlow/Keras results
Problem 3: Iris (Binary Classification)

Binary classification accuracy on test set: 0.90

Problem 4: Iris (Multi-class Classification)

Multi-class classification accuracy on test set: 0.97

Problem 5: House Prices

Mean Absolute Error on test set: 2.39

Problem 6: MNIST

Accuracy on test set: 0.9815

### PyTorch Results (Rewritten from TensorFlow/Keras):
Problem 5: House Prices

Mean Absolute Error on test set: 2.39

Problem 6: MNIST

Accuracy on test set: 0.9708

Comparison:

Performance Metrics:

House Prices (Regression):

TensorFlow/Keras: Mean Absolute Error ≈ 2.39
PyTorch: Mean Absolute Error ≈ 2.39

Both frameworks show similar performance in terms of regression accuracy, indicating that either can be effectively used for such tasks.

MNIST (Image Classification):

TensorFlow/Keras: Accuracy ≈ 0.9815

PyTorch: Accuracy ≈ 0.9708

TensorFlow/Keras slightly outperforms PyTorch in terms of accuracy on the MNIST dataset. 

Ease of Use and Code Readability:

### TensorFlow/Keras

involves fewer lines of code due to its high-level API, which can promote quick prototyping and ease of use.
Well-suited for production deployments and scenarios where rapid development and deployment are critical.

### PyTorch

involves more explicit model construction and training loop setup, which may lead to more lines of code but potentially clearer and more readable implementations, especially for research-oriented tasks.

Offers flexibility with dynamic computation graphs, making it popular among researchers for experimentation and debugging.

## Community and Ecosystem:

### TensorFlow/Keras:

Has a larger community and extensive ecosystem support, including pre-trained models, deployment tools, and integration with other Google Cloud services.

Well-documented with numerous tutorials and resources available.

## PyTorch:

Growing rapidly with strong support from both academia and industry.
Known for its flexibility and intuitive API, particularly favored by researchers for its control over model building and experimentation.